### PR TITLE
Add per-tile flipping and transposing

### LINKED ISSUE #3967: Reimplement ability to flip/rotate/modulate individual tiles in a TileMap (as opposed to creating flipped/rotated/modulated tiles in the TileSet)
### Describe the project you are working on

A top-down procedurally generated rogue-like.

### Describe the problem or limitation you are having in your project

Godot 4.0 Tilemaps no longer support the flipping and transposition of individual tiles. This feature is useful for reusing assets, preventing resource bloat, and synchronizing symmetrical features. The project I'm attempting to test-port from 3.0 to 4.0 relies heavily on this functionality. Because there's no support for this feature, and my maps are built for it, 4.0 isn't currently a viable option for my existing projects.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

The way Godot 3.0 handled flipping and rotation was excellent. I suggest adding that part back into 4.0, which is otherwise superior.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I recommend reinstituting the previously existing buttons back on the icon bar of the tilemap editor. The icons for rotate left, rotate right, flip horizontally, flip vertically, and clear transforms would appear next to the other buttons, or become accessible after clicking on a "transform" button.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This is an editor issue, not a coding issue.

As for work-arounds, not everyone will need this, but for those of us who do, the alternatives are rather painful. It is possible to make an "alternate tile" with these transforms set. This should theoretically work alright for a small number of individual tiles.

Attempting this for an entire tileset, the result is an unusable mess, especially as there is no way to cleanly match an alternate to its corresponding primary tile. It is far less painful being able to make these transformations as needed on the fly in tilemaps rather than in tilesets. You can't necessarily plan in advance what might need to be flipped, it's impractical to flip everything, and in general, the current layout of the alternatives panel leads to bad organization even if you were to try planning this out. Likewise, with the current alternative tiles panel, you lose the functionality to paint "patches" of tiles because they appear out of their intended ordering.

The thought also occurred to me of making a separate layer and using costum dummy tiles to mark where I'd like transformations, but these changes would be non-obvious in the editor, which is required for design best practices.

### Is there a reason why this should be core and not an add-on in the asset library?

Mostly, because this is re-implementing a 3.0 feature that got lost in the shuffle. In my opinion, an add-on should not be necessary for porting projects to 4.0.

### PR DESCRIPTION
**Pull request by KoBeWi**
Closes https://github.com/godotengine/godot-proposals/issues/3967

https://github.com/godotengine/godot/assets/2223172/0a4e4a41-f449-4d6f-9a9e-c630139d0bae

https://github.com/godotengine/godot/assets/2223172/837f3ba2-74c2-472d-af98-d0b99c0f452b

### CODE REVIEW

**Comment by KoBeWi** [comment]
Finished:

https://github.com/godotengine/godot/assets/2223172/0a4e4a41-f449-4d6f-9a9e-c630139d0bae

Although I still need to do some cleanup.

Also, since transform is stored as part of alternate id now, the maximum number of alternate tiles per tile is now more limited (from 32768 to 4096). We should consider enforcing this limitation or documenting it.

**Comment by KoBeWi** [comment]
Pattern rotating:

https://github.com/godotengine/godot/assets/2223172/837f3ba2-74c2-472d-af98-d0b99c0f452b

**Comment by KoBeWi** [comment]
Rebased and squashed (there was no reason to keep the history after the recent rework).
I also fixed the shortcuts now that #80317 was merged (it needed a small tweak though).

**Comment by groud** [comment]
Sooo, looks mostly good. I wonder about the following situations though:
- what to do with half-offset layouts (isometric, half-offset square and hexagonal).
- what to do when the tiles are not square.

Maybe we might want to implement a way to disable the rotation buttons when height != width, and only allow flips ? That would avoid some confusion I believe. Though, for special cases, I guess we could have a way to "force rotations" anyway. Like maybe a checkbox or something.

**Comment by KoBeWi** [comment]
> Maybe we might want to implement a way to disable the rotation buttons when height != width, and only allow flips ?

Big tiles seem to work fine:

https://github.com/godotengine/godot/assets/2223172/fd961ab1-4973-4f9d-8b9c-ed1edb77099d

Not sure what confusion are you referring to.

As for non-square tilesets, I don't have one to test how it works. Half-offset square doesn't seem to have problems with transforms. Hexagon tiles will probably look wrong when rotated, but I doubt anyone expects them to retain shape after rotating. Isometric tiles will probably be fine.

**Comment by groud** [comment]
> Big tiles seem to work fine:
> Not sure what confusion are you referring to.

Try with a pattern ;)

> As for non-square tilesets, I don't have one to test how it works. Half-offset square doesn't seem to have problems with transforms. Hexagon tiles will probably look wrong when rotated, but I doubt anyone expects them to retain shape after rotating. Isometric tiles will probably be fine.

Same, try with a pattern there.

Maybe the solution is to disable only for patterns and not single tiles there ? That would probably make sense I think.

**Comment by KoBeWi** [comment]
Still looks correct:

https://github.com/godotengine/godot/assets/2223172/81c9affb-b014-420e-822c-8a8609ba5326

Half-offset squares rotate a bit weird, but they correctly fit into the grid:

https://github.com/godotengine/godot/assets/2223172/62febff6-6185-4bc9-b812-cc7349451c70

It doesn't break anything and produces a result expected from the given grid shape.

**Comment by groud** [comment]
> Half-offset squares rotate a bit weird, but they correctly fit into the grid:

Well, yeah, that's what I mean. If you have a more meaningful pattern (like you usually do) with a house or something, the rotation does not make any sense. I think the behavior in the second video should likely be not possible.

**Comment by KoBeWi** [comment]
Done.

https://github.com/godotengine/godot/assets/2223172/00287669-eb1a-4561-8cf6-78a493130da6

**Comment by KoBeWi** [comment]
Rebased after #81070
Not sure when to clear polygon cache now. Might need separate cache for every polygon type :/

**Review by groud — APPROVED** [review]
I think this looks good overall. The only thing I am not sure about is whether or not the polygon cache should be in the TileMap itself or managed by the TileSet/TileSetAtlasSource itself. Just in case you might have several TileMaps sharing the same TileSet, it might be a bit more efficient.

File: editor/plugins/tiles/tile_map_editor.cpp [hunk 1]
```diff
@@ -924,18 +950,18 @@ void TileMapEditorTilesPlugin::forward_canvas_draw_over_viewport(Control *p_over
 							Rect2 dest_rect;
 							dest_rect.size = source_rect.size;
 
-							bool transpose = tile_data->get_transpose();
+							bool transpose = tile_data->get_transpose() ^ bool(E.value.alternative_tile & TileSetAtlasSource::TRANSFORM_TRANSPOSE);
```
[the remark this answers is not in the document]
Comment (KoBeWi) [reply editor/plugins/tiles/tile_map_editor.cpp:953]: There's no non-bitwise equivalent of `^` AFAIK. I did make sure it works as expected.

[the remark this answers is not in the document]
Comment (groud) [reply editor/plugins/tiles/tile_map_editor.cpp:953]: > There's no non-bitwise equivalent of ^ AFAIK. I did make sure it works as expected.

Well, as I mentioned, it's `!=`. But that's fine anyway, if you checked this was working correctly, we can keep it that way, it's more understandable IMO.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**